source/gridfs/gridfs-spec.md

```diff
@@ -0,0 +1,1021 @@
+as the id for the file being uploaded. Otherwise the application provides the value. Drivers store the contents of the
+user file in the chunks collection by breaking up the contents into chunks of size "chunkSizeBytes". For a non-empty
+user file, for each n<sup>th</sup> section of the file, drivers create a chunk document and set its fields as follows:
+
+- Files_id: the id generated for this stored file.
+- N: this is the n<sup>th</sup> section of the stored file, zero based.
+- Data: a section of file data, stored as BSON binary data with subtype 0x00. All chunks except the last one must be
+  exactly 'chunkSizeBytes' long. The last chunk can be smaller, and should only be as large as necessary.
+
+Historically, while streaming the user file, drivers computed an MD5 digest for the (now deprecated) 'md5' field of the
+files collection document. If drivers preserve this behavior for backwards compatibility, they MUST provide the
+'disableMD5' member of GridFSBucketOptions. When 'disableMD5' is true, drivers MUST NOT compute an MD5 digest or include
+it in the files collection document. If drivers no longer support the deprecated 'md5' field, they MUST NOT provide the
+'disableMD5' member (or it MUST be a no-op) and MUST NOT compute MD5.
+
+After storing all chunk documents generated for the user file in the "chunks" collection, drivers create a files
+collection document for the file and store it in the files collection. The fields in the files collection document are
+set as follows:
+
+- Length: the length of this stored file, in bytes.
+- Chunksize: the chunk size in bytes used to break the user file into chunks. While the configuration option is named
+  "chunkSizeBytes" for clarity, for legacy reasons, the files collection document uses only "chunkSize".
+- Uploaddate: a BSON datetime object for the current time, in UTC, when the files collection document was created.
+- Md5: MD5 checksum for this user file, computed from the file"s data, stored as a hex string, if computed, otherwise
+  omitted.
+- Filename: the filename passed to this function, UTF-8 encoded.
+- Contenttype: the "contentType" passed in the options, if provided; otherwise omitted.
+- Aliases: the array passed in the options, if provided; otherwise omitted.
+- Metadata: the "metadata" document passed in the options, if provided; otherwise omitted.
+
+If a user file contains no data, drivers MUST still create a files collection document for it with length set to zero.
+Drivers MUST NOT create any empty chunks for this file.
+
+Drivers MUST NOT run the 'filemd5' database command to confirm that all chunks were successfully uploaded. We assume
+that if none of the inserts failed then the chunks must have been successfully inserted, and running the 'filemd5'
+command would just be unnecessary overhead and doesn't work for sharded chunk keys anyway.
+
+**Operation Failure**
```

Comment:
Various `**foo**` lines should also be level 4 headings. Not sure why these were never titles in the RST, but this is a good time to fix that.

Reply:
Done